### PR TITLE
feat: allow prefiltering to be used with an index

### PR DESCRIPTION
**Pull request by westonpace**
Currently, prefiltering is restricted to searches that do not use an index.  This PR enables prefiltering to be used on an indexed column.  The approach is to first run a query to figure out the valid row ids and then pass this list of valid row ids to the index.

### CODE REVIEW

File: rust/lance-core/src/datatypes/schema.rs [hunk 1]
```diff
@@ -44,8 +44,7 @@ struct SchemaFieldIterPreOrder<'a> {
 impl<'a> SchemaFieldIterPreOrder<'a> {
     #[allow(dead_code)]
     fn new(schema: &'a Schema) -> Self {
-        let mut field_stack = Vec::new();
-        field_stack.reserve(schema.fields.len() * 2);
+        let mut field_stack = Vec::with_capacity(schema.fields.len() * 2);
```
Comment (westonpace) [inline rust/lance-core/src/datatypes/schema.rs:47]: This is unrelated but clippy keeps warning about it on my system so I just threw it in

File: python/python/benchmarks/test_search.py [hunk 2]
```diff
@@ -168,3 +169,28 @@ def test_ivf_pq_index_search(test_dataset, benchmark):
         ),
     )
     assert result.num_rows > 0
+
+
+@pytest.mark.benchmark(group="query_ann")
+@pytest.mark.parametrize(
+    "selectivity, prefilter, use_index",
+    list(itertools.product((0.25, 0.75), (False, True), (False, True))),
+)
```
Comment (wjones127) [inline python/python/benchmarks/test_search.py:178]: I think if you do this, it's a little more readable and also pytest might generate easier to understand test names:
```suggestion
@pytest.mark.parametrize("selectivity", (0.25, 0.75))
@pytest.mark.parametrize("prefilter", (False, True))
@pytest.mark.parametrize("use_index", (False, True))
```

File: rust/lance/src/dataset/scanner.rs [hunk 3]
```diff
@@ -1765,6 +1782,80 @@ mod test {
...
+            .unwrap();
+        assert_eq!(batches.len(), 1);
+
+        let first_match = batches[0]
+            .column_by_name(ROW_ID)
+            .unwrap()
+            .as_any()
+            .downcast_ref::<UInt64Array>()
+            .unwrap()
+            .values()[0];
```
Comment (wjones127) [inline rust/lance/src/dataset/scanner.rs:1854]: FYI, if you don't need to handle the errors, there are some shorthands you can use here:
```suggestion
        let first_match = batches[0][ROW_ID].as_primitive::<UInt64type>()
            .values()[0];
```

File: rust/lance/src/index/prefilter.rs [hunk 4]
```diff
@@ -149,12 +179,40 @@ impl PreFilter {
...
-            .collect::<Vec<_>>()
+        let enumerated_ids = row_ids.iter().enumerate();
+        match (&self.block_list, &self.allow_list) {
+            (Some(block_list), Some(allow_list)) => {
+                // Only take rows that are both in the allow list and not in the block list
+                let block_list = block_list.get_ready();
+                let allow_list = allow_list.get_ready();
+                enumerated_ids
+                    .filter(|(_, row_id)| {
+                        !block_list.contains(**row_id) && allow_list.contains(**row_id)
```
Comment (wjones127) [inline rust/lance/src/index/prefilter.rs:190]: I wonder: would the `allow_list` filter be typically more selective? And if so, should we do that first?

File: rust/lance/src/index/prefilter.rs [hunk 5]
```diff
@@ -33,17 +34,29 @@ use crate::format::RowAddress;
...
 pub struct PreFilter {
-    dataset: Arc<Dataset>,
-    has_deletion_vectors: bool,
-    has_missing_fragments: bool,
-    block_list: Arc<SharedPrerequisite<Arc<RoaringTreemap>>>,
+    // Expressing these as tasks allows us to start calculating the block list
+    // and allow list at the same time we start searching the query.  We will await
+    // these tasks only when we've done as much work as we can without them.
+    block_list: Option<Arc<SharedPrerequisite<Arc<RoaringTreemap>>>>,
+    allow_list: Option<Arc<SharedPrerequisite<Arc<RoaringTreemap>>>>,
```
Comment (wjones127) [inline rust/lance/src/index/prefilter.rs:51]: I really like this approach with `SharedPrerequisite` btw 🙌

**Review by wjones127 — APPROVED** [review]
Looks good. Left a few information comments, but IMO this is good to go as-is.

File: rust/lance/src/dataset/scanner.rs (hunk 3 above)
Comment (westonpace) [reply rust/lance/src/dataset/scanner.rs]: Thanks.  Switched to these.

File: rust/lance/src/index/prefilter.rs (hunk 4 above)
Comment (westonpace) [reply rust/lance/src/index/prefilter.rs:190]: I went ahead and put it first.  We can measure this in more detail at some future time.

**Pull request merged**